biggus/__init__.py

```diff
@@ -2781,71 +2826,162 @@ def ndarray(self):
-
-    """
-    return _Elementwise(a, b, np.add, np.ma.add)
-
-
-def sub(a, b):
-    """
-    Return the elementwise evaluation of `a - b` as another Array.
-
-    """
-    return _Elementwise(a, b, np.subtract, np.ma.subtract)
-
-
-def multiply(a, b):
-    """
-    Return the elementwise evaluation of `a * b` as another Array.
-
-    """
-    return _Elementwise(a, b, np.multiply, np.ma.multiply)
-
-
-def floor_divide(a, b):
-    """
-    Return the elementwise evaluation of `a // b` as another Array.
-
-    """
-    return _Elementwise(a, b, np.floor_divide, np.ma.floor_divide)
-
-
-def divide(a, b):
-    """
-    Return the elementwise evaluation of 'a / b' as another Array.
-
-    """
-    return _Elementwise(a, b, np.divide, np.ma.divide)
-
-
-def true_divide(a, b):
-    """
-    Return the elementwise evaluation of ``np.true_divide`` as another Array.
-
-    """
-    return _Elementwise(a, b, np.true_divide, np.ma.true_divide)
-
-
-def power(a, b):
-    """
-    Return the elementwise evaluation of `a ** b` as another Array.
-
-    """
-    return _Elementwise(a, b, np.power, np.ma.power)
+def _unary_fn_wrapper(name, function_to_wrap, masked_equivalent=None,
+                      fn_name=None):
+    @functools.wraps(function_to_wrap, assigned=('__name__', '__doc__'))
+    def wrapped_function(a):
+        return _Elementwise(a, None, function_to_wrap, masked_equivalent)
+    doc_str = ("Return the elementwise evaluation of {}(a) "
+               "as another Array.".format(name))
+    wrapped_function.__name__ = fn_name or function_to_wrap.__name__
+    wrapped_function.__doc__ = doc_str
+    return wrapped_function
+
+
+def _dual_input_fn_wrapper(name, function_to_wrap, masked_equivalent=None,
+                           fn_name=None):
+    @functools.wraps(function_to_wrap, assigned=('__name__', '__doc__'))
+    def wrapped_function(a, b):
+        return _Elementwise(a, b, function_to_wrap,
+                            masked_equivalent)
+    doc_str = ("Return the elementwise evaluation of {}(a, b) "
+               "as another Array.".format(name))
+    wrapped_function.__name__ = fn_name or function_to_wrap.__name__
+    wrapped_function.__doc__ = doc_str
+    return wrapped_function
+
+
+def _ufunc_wrapper(ufunc):
+    """
+    A function to generate the top level biggus ufunc wrappers.
+
+    """
+    if not isinstance(ufunc, np.ufunc):
+        raise TypeError('{} is not a ufunc'.format(ufunc))
+
+    name = ufunc.__name__
+    # Get hold of the masked array equivalent, if it exists.
+    ma_ufunc = getattr(np.ma, name, None)
+    if ufunc.nin == 2 and ufunc.nout == 1:
+        func = _dual_input_fn_wrapper('np.{}'.format(name), ufunc, ma_ufunc,
+                                      name)
+    elif ufunc.nin == 1 and ufunc.nout == 1:
+        func = _unary_fn_wrapper('np.{}'.format(name), ufunc, ma_ufunc,
+                                 name)
+    else:
+        raise ValueError('Unsupported ufunc (nin {}, nout {})'
```

Comment:
I wonder if this error message could be a little friendly:
```python
'Unsupported ufunc (nin 5, nout 3)'
```
Does read a little like a football score :soccer: 

How about:
`'Unsupported ufunc {!r} with {} input arrays & {} output arrays.'.format(name, ufunc.nin, ufunc.nout)`

Thus:

```python
'Unsupported ufunc "sqrt" with 5 input arrays & 3 output arrays.'
```
?